### PR TITLE
Count/find imaged moments by images and association link name

### PR DESCRIPTION
**Pull request by kevinsbarnard**
Duplicate of https://github.com/kevinsbarnard/annosaurus/pull/1.

Added endpoints:
1. `/v1/imagedmoments/count/all` returns a count of all imaged moments
2. `/v1/imagedmoments/count/images` returns a count of imaged moments with at least one valid (non-null URL) image reference
3. `/v1/imagedmoments/find/images` returns the imaged moments " " " "
4. `/v1/imagedmoments/count/linkname/:linkname` returns a count of imaged moments with at least one association with the given link name
5. `/v1/imagedmoments/find/linkname/:linkname` returns the imaged moments " " " "

Unit tests to cover each new endpoint are included.

### CODE REVIEW

File: src/main/scala/org/mbari/vars/annotation/dao/jpa/ImagedMomentImpl.scala [hunk 1]
```diff
@@ -115,6 +115,23 @@ import scala.collection.mutable
...
       name = "ImagedMoment.countByVideoReferenceUUID",
       query = "SELECT COUNT(*) FROM imaged_moments WHERE video_reference_uuid = ?1"
+    ),
+    new NamedNativeQuery(
+      name = "ImagedMoment.countAll",
+      query = "SELECT COUNT(*) FROM imaged_moments"
+    ),
+    new NamedNativeQuery(
+      name = "ImagedMoment.countWithImages",
+      query = "SELECT COUNT(*) FROM imaged_moments i " +
```
Comment (hohonuuli) [inline src/main/scala/org/mbari/vars/annotation/dao/jpa/ImagedMomentImpl.scala:125]: Change to `COUNT(DISTINCT i.uuid)`. Other wise you get a count of sum(imagedMoment * it's image references)

File: src/main/scala/org/mbari/vars/annotation/dao/jpa/ImagedMomentImpl.scala [hunk 2]
```diff
@@ -115,6 +115,23 @@ import scala.collection.mutable
...
+    ),
+    new NamedNativeQuery(
+      name = "ImagedMoment.countWithImages",
+      query = "SELECT COUNT(*) FROM imaged_moments i " + 
+        "INNER JOIN image_references ir ON ir.imaged_moment_uuid = i.uuid " +
+        "WHERE ir.url IS NOT NULL"
+    ),
+    new NamedNativeQuery(
+      name = "ImagedMoment.countByLinkName",
+      query = "SELECT COUNT(*) FROM imaged_moments i " +
```
Comment (hohonuuli) [inline src/main/scala/org/mbari/vars/annotation/dao/jpa/ImagedMomentImpl.scala:131]: Change to `COUNT(DISTINCT i.uuid)`

File: src/test/scala/org/mbari/vars/annotation/api/v1/ImagedMomentV1ApiSpec.scala [hunk 3]
```diff
@@ -153,4 +162,104 @@ class ImagedMomentV1ApiSpec extends WebApiStack {
     }
   }
 
+  it should "count and find with images" in {
```
Comment (hohonuuli) [inline src/test/scala/org/mbari/vars/annotation/api/v1/ImagedMomentV1ApiSpec.scala:165]: Modify the test to add two images at the same imagedMoment to verify the count returned is 1.

**Review by hohonuuli — CHANGES_REQUESTED** [review]
@kevinsbarnard Looks good. Just minor changes need to the count queries.

File: src/main/scala/org/mbari/vars/annotation/dao/jpa/ImagedMomentImpl.scala (hunk 1 above)
Comment (hohonuuli) [reply src/main/scala/org/mbari/vars/annotation/dao/jpa/ImagedMomentImpl.scala:125]: For example: as written returns 1,154,916. Using distinct returns 606,273 (i.e. There's a png and jpg on each imagemoment)

File: src/test/scala/org/mbari/vars/annotation/api/v1/ImagedMomentV1ApiSpec.scala (hunk 3 above)
Comment (kevinsbarnard) [reply src/test/scala/org/mbari/vars/annotation/api/v1/ImagedMomentV1ApiSpec.scala:165]: Test now creates 10 images at the same time, then asserts only one imaged moment exists with the same number of images.

File: src/main/scala/org/mbari/vars/annotation/dao/jpa/ImagedMomentImpl.scala (hunk 2 above)
Comment (kevinsbarnard) [reply src/main/scala/org/mbari/vars/annotation/dao/jpa/ImagedMomentImpl.scala]: Done

File: src/main/scala/org/mbari/vars/annotation/dao/jpa/ImagedMomentImpl.scala (hunk 1 above)
Comment (kevinsbarnard) [reply src/main/scala/org/mbari/vars/annotation/dao/jpa/ImagedMomentImpl.scala]: Done

**Review by hohonuuli — APPROVED** [review]

**Comment by hohonuuli** [comment]
Running integration tests before merging into master.

**Comment by hohonuuli** [comment]
Integration tests are good. Nice work! Just a heads up that, as a I mentioned, the JPA fetch is sluggish so if it turns out to be a show stopper for your use case, I can get you going on a SQL alternative. The changes are in release at https://github.com/mbari-media-management/annosaurus/releases/tag/0.13.0 . New arm/amd images pushed to docker hub.

**Comment by kevinsbarnard** [comment]
Thanks Brian! I see what you mean- I'll see what I can do next week.

**Pull request merged**